### ISSUE TITLE
Version 1.7.0. missing texts

### ISSUE THREAD

**Issue by FASTNAVi** [NPBruce/valkyrie#758]
Hey, I just download and apply your new Valkyria versions. It is great, but I lost a plenty of text of my tokens and events. And I mean really plenty of them, about 80%. I had almost scenario done so I am mad right now and looking for some solution. Also a lot of buttons change name. Instead of "continue" is there name of token/event.button1 etc. It looks like the all missing texts have also the problem with name of button. 

Please, just tell me that it is only small problem and I can fix it easy..

**Comment by scrubbless** [comment]
I noticed something similar, this seems to happen when text is copied in.
If you copy some text in, hit save. Sometimes part of the text will dissappear. Its very confusing.

I also noticed that sometimes when you create a new event, then rename it. It loses the link (and puts the event name in the button) like FASTNAVi is mentioning above.

One thing I noticed is the events were still there, so you could re-add them under the button (the text in the button giving you a clue).

**Comment by FASTNAVi** [comment]
Well, my problem is different. I didn't copy texts. I was writing most of text directly in Valkyria. Yesterday they all were there, today not. And I mean ALL text. About 80% events and tokens have empty text box and bad name of buttons.. I have a dozens hours in it and I am afraid I lost it :-( At least the events logic is still working.. but texts are gone.

**Comment by scrubbless** [comment]
Do you have a previous version you can roll back to? 

That sounds bad FASTNAVi, i feel for you.

**Comment by FASTNAVi** [comment]
Well, I am afraid, I have a plan to do backup directly after finishing last details and before starting testing and translating into english. I was hoping, that this will be in sunday.. this sunday.. my last recovery is from time, when I was on beginning.. well my fault..

**Please, everyone, make copies of your work every few hours.** - maybe this advice is good for beginners like me. You know, in MoM tutorial pdf file..

Thank you scrubbless.. I saw localisation file.. it looks like the problem is there, because all events logic and etc are intact. So just one file si bad.. but still.. uuuaaaa =(

**Comment by scrubbless** [comment]
Have you checked the localisation file for any of your text? Sometimes i've noticed that lost text can sometimes appear in here and just happens to be un-linked to the events.

**Comment by FASTNAVi** [comment]
Well I see problem now. My localization file is much more smaller than last time. And I look at all inputs. I have all texts made before 14.1. there, but from the next day and futher I have nothing. My loc file is ending with text made in 14.1. (one day after last backup so I know that time very well).

**Comment by FASTNAVi** [comment]
I suppose, my fault was to update valkyria without backup my scenario.. it is certainly my fault :( I am going out for some walk and calm down.. right now I am sooo sad.

**Comment by scrubbless** [comment]
I use github.

Its a great way to track changes and regularly back up.

**Comment by NPBruce** [comment]
Sorry this has happened.  Can you please let me know what version you were running before and after the issue, so that I can try to stop this?

Additionally I do recommend learning and using version control, not just for this but for all computer work it is a great tool to have.

**Comment by NPBruce** [comment]
PS: The 1.7 series (as per title) are currently all in prerelease.  While I try not to introduce any serious bugs they do not have the same level of testing as the stable releases, I only recommend updating to a prerelease because there is a specific feature/fix you want or to help me out with development/testing.  In either case, use caution.

**Comment by scrubbless** [comment]
Ive only been using 1.7.0b because of the features that make it stand out from 1.6.1.

I version control constantly since the image issue. But id rather risk losing a bit of data to identify these bugs and get 1.7 stable!

**Comment by FASTNAVi** [comment]
NPBruce: I had 1.6.1. Now I have version which you release (or update) yesterday. It is 1.7.0c. After update, I also use "reimport content" for MoM. Then I launch my scenario in editor and I was surprise. First I really like sorting things in components, then I realize the problem. I understand that is prerelease but I don't expect that. But missing texts and buttons are still better than complete missing scenario :D The most important thing - the quest logic - is still present. I just need to do all story once more again.

**Comment by scrubbless** [comment]
Hi @npbruce, i wanted to tag you bause this is a long thread.

I have observed rhe same issue. It seems that 1.7.0c is not compatible with 1.7.0b created events.

Content created in 1.6.1 shows. But content created in1.7.0b has no dialog.
![screenshot_2018-01-31-20-10-23](https://user-images.githubusercontent.com/32058089/35645571-263f3416-06c4-11e8-9cce-cdc4adaceed3.png)

Not usre if this is a problem wwoth 1.7.0c or an undiscovered issue in 1.7.0b.

**Comment by FASTNAVi** [comment]
Well, this was exactly my problem..

Scrubbless: Did you lost some data or you was wise and have backup? :)

**Comment by scrubbless** [comment]
I have multiple versions of my scenarios. However, my latest is only in 1.7.0b.

It would be a LOT of copy and pasting if i had to "rewrite it" in 1.7.0c (or 2.0 if the problem carries over).

I would highly recommend using github. You can work locally on you scenario and every so often save it, commit it and push it to the internet. All without even closing Valkyrie. 

I work on several machines so it means i can switch to any machine, anywhere that has a new connection and get the latest copy. It also means i can revert changes that break my scenario.

In this case i didnt commit and push anything back. So i could just clone the current (before corruption) version back to my machine.

**Comment by FASTNAVi** [comment]
Hi guys! My scenario is almost done. I am just testing it. But I have still the same issue. When I hit SAVE in the editor, I have a lottery if texts disappear or not. It is really annoying. I don't know where is 
the problem. It looks like that I have about 25% probability of disappearing texts. And about 50/50 chance to have one of two different broken txt files.

I can put here my scenario just for you and also saved files after the disappearing texts, if it helps you to identify the problem.

**Comment by scrubbless** [comment]
I found a work around for this in #757

**Comment by FASTNAVi** [comment]
Thanks scrubbless, you are right about that. Also it helps me converting the localization file from Windows format to UNIX format in my notepad++. And I think, the problem was also when I start event text with quotation marks. I simply add a space in front of it and the saving in editor looks like working fine now.

**Comment by scrubbless** [comment]
Hey Fastnavi, you may want to keep this issue open to track the 1.7.0c "corrupting scenario's" issue above.

**Comment by redwolf2** [comment]
I have seen that happen, when there was an invalid quotation. Like "hallo " world"
https://github.com/NPBruce/valkyrie/issues/575

This may help to solve the problem:
https://github.com/NPBruce/valkyrie/pull/789#commits-pushed-bff66f6

However, the problem seem to lay in the save logic of the scenario. If only one event text causes a conflict, all other event texts after that go missing. There should be a better fail mechanism, that continues to write the rest of the content, when shit hits the fan. We may also add a test suite for DictionaryI18n.

**Comment by scrubbless** [comment]
Thanks for the update Redwolf2.

I wonder if this issue is linked to the copy and paste error. I noticed phantom text in the localisation file that had since been deleted from events. Checking GitHub previous versions it appears that the text was old stuff that I had pasted in and had disappeared when saving.

My latest scenario doesn't have any issues in 1.7.0c, but I also haven't had the lost text when pasting (I'm wise to it now!)

**Comment by redwolf2** [comment]
Here is how to reproduce it.
New Scenario. Add Event0 with the text `"hallo " world"`

Press Menu > Save

```
.,English
quest.name,Szenario 1
quest.description,Szenario 1...
Event0.text,"""hallo "" world"""
Event0.button1,Continue
```

(Event text turns empty in UI)

Press Menu > Save

```
.,English
quest.name,Szenario 1
quest.description,Szenario 1...
```

**Comment by NPBruce** [comment]
@redwolf2 I couldn't produce the problem following the steps you detail here. Does anyone have an example of data that works in 1.7.0b but not 1.7.0c? Or is the problem that certain operations in 1.7.0c lose data? Still not exactly clear on what is happening, so I want to be sure I am fixing the right thing.

**Comment by redwolf2** [comment]
I reproduced the problem with a build made from the latest "development" source. I claimed that it will also happen on 1.7.0c. I will test it again with 1.7.0c, just to be sure.

**Comment by scrubbless** [comment]
The problem presents in Stress and Strain. Towards the end.

I'll create a way to skip to it and upload it when I got to a computer later. Its near the end after entering the holding cell (basement)

**Comment by NPBruce** [comment]
I think there are two problems here.  One is a saving issue that @redwolf2 has addressed, but there may be another issue if existing scenarios are not working correctly, which I have not got to the bottom of.

**Comment by NPBruce** [comment]
I have addressed the issue for saving scenarios, but it is possible that existing scenarios have bad entries which makes parsing a problem.  It may be that these can't be fixed automatically (requiring scenario edit), but if specific examples are shown I will investigate.

**Comment by scrubbless** [comment]
I am still having the same issue:
![image](https://user-images.githubusercontent.com/32058089/38997407-417daf42-43e5-11e8-8405-067db23370d0.png)

![image](https://user-images.githubusercontent.com/32058089/38997415-48711a32-43e5-11e8-9292-669420d15fda.png)


This is in the official "Stress and Strain" that you can download from the scenario's section. Just select the explore tokens until you get to the plaza, then kill the riot and enter the Sheriff's office to the north. The right hand explore token is blank, and then the search tokens in the room have the "xxx.button1" text.

This was saved in 1.6.1, loading it in 1.7.0d seems to delete a chunk of text from various events.

Here are the events for comparison:

1.6.1
![image](https://user-images.githubusercontent.com/32058089/38997913-69c86a22-43e6-11e8-9d30-9eac5defcaad.png)

Haha thats odd, importing content with the updated Official app now changes up all the icon images in 1.6.1.

1.7.0d
![image](https://user-images.githubusercontent.com/32058089/38997842-49c5e1aa-43e6-11e8-9f37-91ab05955a16.png)

**Comment by scrubbless** [comment]
A scenario I created in 1.7.0b seems fine, I don't see any of the "xxx.button1" errors.
This is going to be a problem for people trying to maintain scenarios created in 1.6.1 and prior.

**Comment by redwolf2** [comment]
I tested 1.7.0d and can confirm it is working

**Comment by scrubbless** [comment]
The scenario I created from scratch in 1.7.0b is working compeltely fine. It just seems to have problems with backwards compatibility to 1.6.1.

1.7.0b works completely fine with the scenario created in 1.6.1, but 1.7.0c - 1.7.0d seem to delete the text. Not sure why, but it has something to do with a change in 1.7.0c.

I suspect its just a subtle but in 1.6.1 that was presenting as working, but has broken with whatever changes happened in 1.7.0c.

Like making a web page - just because it test OK in IE, doesn't mean its not a bug ridden mess of code! ;D

**Comment by scrubbless** [comment]
Its not the end of the world, if its just an error in my scenario. I can load up a version in 1.6.1, find the events that are broken in 1.7.0d and copy the text over. It doesn't seem to break the event logic.

However this is all contingent on one of the pre-releases being promoted to the latest stable version.

**Comment by redwolf2** [comment]
If you can copy the text, that broke your scenario from the `Localization.English.txt` here. Then we can investigate. You can also post this scenario here for analysis.

**Comment by scrubbless** [comment]
Will do, will probably be tomorrow now.

Though it is in the download list, so you could copy it.

I'll dig out the localisation text if possible.

**Comment by NPBruce** [comment]
OK I think I know how to support existing text without change, but the more examples I can get of the fault the more likely I will get it right.

**Comment by NPBruce** [comment]
Here is an example of a problem I have found:

<pre>
EventSheriffPersuadeFail_Tile15.text,As I said, I can't leave here. I have nothing else to offer you" he replies, you notice a slight look of guilt in his eye
EventSheriffPersuadeFail_Tile15.button1,Continue
</pre>

The single quote will result in EventSheriffPersuadeFail_Tile15.text including the new line and all following text until it hits a line with an odd number of " characters.  In this case I can fix it by only doing this if a " character is used immediately after EventSheriffPersuadeFail_Tile15.text, but there may still be cases which start with a " and have an odd number which are not multi line.

Another option would be to ensure that when in multi line text, line must not have a , before the first space to detect run on into the next entry, but this is very hacky and likely to break ligitimate uses.

TL;DR: The more examples of the fault I can find, the better the solution.

**Comment by scrubbless** [comment]
Hi Bruce,

The leading " kept being removed which was a bug in 1.6.1. since then I stopped using quotes and went to italics for speak text.

If this is the cause I can amend the text. Can I do this directly in the localisation.english file?

**Comment by NPBruce** [comment]
Sure, you should be able to use quotes and newlines without issue going forward, but I would like to get 1.7 correctly loading existing scenarios, so that I don't break anything already out there.  If you are going to update in a way that doesn't work in 1.6 you should update the quest format to 9 (which means noone will be able to use it in 1.6).

**Comment by scrubbless** [comment]
I have a version 3 that is saved as version 9, I was just going to remove all " and replace speak text in <i>.

> so that I don't break anything already out there.

FFG did that with the latest Official App update, none of the icons are displaying correctly in 1.6.1, even in older scenario's #798

**Comment by NPBruce** [comment]
Yep, but when updating to 1.7 I would like all existing scenarios to work.

**Comment by NPBruce** [comment]
All cases I am aware of should be fixed for 2.0 release.

**Comment by scrubbless** [comment]
Hi @NPBruce , v2.0 of Valkyrie doubled down on these issues in my scenario. But I have fixed them all in the Localisation file. (I think!)

You (or was it @redwolf2 ?) were partly right about the quotation marks all over the place. It seems that every instance of a quotation mark "This is Dialogue from a person" seemed to add  additional quotation marks all over the place. 

I used find and replace in the localisation file to change the "Speech" into <i>Speech</i> I also removed phantom solo quotation marks causing havoc.

The other part of the issue for me at least causing the buttons to not display and text to be missing was due to Events/Tokens and Spawns having the carriage returns before them translated to \n.

So instead of each event (etc) starting on a new line, multiple events shared the same line.

I also cleared these out using find and replace but here is an example:
![image](https://user-images.githubusercontent.com/32058089/39185832-64e6e1fe-47c0-11e8-90a1-bf75e1957f25.png)


I am not sure if this is of any help, just wanted to let you know if something similar crops up.

**Comment by NPBruce** [comment]
Can you please give any examples of text that added additional quotation marks?  And is this when playing, opening in the editor or saving?

**Comment by NPBruce** [comment]
To clarify, any text with " characters will now look like this in the data when saved:

foo "bar" thing -> "foo ""bar"" thing"

The whole thing will be in " and each actual quote will be doubled.  This allows quotes at the start and end ("""foo""bar"/"foo""bar""") without losing characters and is compatible with FFG newline entries:

"foo
bar"

You could use this format if you like, but resaving will turn this into "foo\nbar".

If the first character in the string is not ", then it will be left as-is and only a single line will be read.

**Comment by scrubbless** [comment]
I believe thats what it looked like.

I fired up the older version of the scenario (2.5) that was created in 1.6.1 and edited in 1.7.0b:

It works fine in 1.6.1 and 1.7.0b. But when I run it in 2.0 (or 1.7.0b and above) I get errors like this:
![image](https://user-images.githubusercontent.com/32058089/39235460-d8d7bd74-486d-11e8-94fe-0db7423da341.png)

I've attached the scenario files, there is even more missing text in 2.0 than there was in 1.7.0b.

Here is the file:

[StressAndStrainV2.5.zip](https://github.com/NPBruce/valkyrie/files/1946224/StressAndStrainV2.5.zip)


> So instead of each event (etc) starting on a new line, multiple events shared the same line.

As for the \nEventName issue, I think I may have created that myself when I used find and replace to clear out the quotation marks. So that is probably a red herring.

**Comment by scrubbless** [comment]
That is actually text that I copied in way back and disappeared after I hit save. 
It was the very first intro text before I had full decided on a town name!

I am not surprised it caused issues, I found it too and deleted it when I cleaned up the localisation file for v3.

The text was complete when I copied it in, but the save bug wiped half of it out... well since it was in the text it just hid it from the editor and scenario!

**Comment by NPBruce** [comment]
I have decided to close this rather than complicate the existing code.  I am not aware of existing scenarios that have this problem, and future scenarios should not be exposed to this problem.

**Comment by scrubbless** [comment]
Yeh I think its a bug in a previous version that has since been fixed (copy and paste) thats remained hidden for multiple versions and has just cropped up

Its not worth wasting time on, since no-one else has seen it and I have cleaned up the code in question.

Thanks for looking though @NPBruce !
